File: sls_calib/marker_detector.py
"""
Marker Detector - Python port of SLSMarkerDetector from markerdetector.cpp

Finds circular markers in images using contour analysis with subpixel refinement.
"""

import math
import time
from typing import List, Tuple

import cv2
import numpy as np

# Type alias: a marker is ((cx, cy), area)
Marker = Tuple[Tuple[float, float], float]


class SLSMarkerDetector:
    """Detects circular markers in images via contour fitting and subpixel refinement."""

    def __init__(self) -> None:
        pass

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def detectMarkers(
        self,
        image: np.ndarray,
        smooth: bool = False,
        debug: bool = False,
    ) -> Tuple[List[Marker], str]:
        """
        Detect circular markers in an image.

        Args:
            image:  Input image (BGR or grayscale numpy array).
            smooth: Apply GaussianBlur before processing.
            debug:  Write debug images (mask.png, contour.png, circle.png).

        Returns:
            (markers, error_info) where *markers* is a list of
            ``((cx, cy), area)`` tuples and *error_info* is an empty
            string on success or an error description.
        """
        markers: List[Marker] = []

        if image is None or image.size == 0:
            return markers, "标志点图像为空！\n"

        t0 = time.perf_counter()
        error_info = ""

        try:
            if image.ndim == 3 and image.shape[2] == 3:
                gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
            else:
                gray = image
            self._findCircles(markers, gray, smooth, debug)
        except cv2.error as exc:
            error_info = str(exc)

        elapsed = (time.perf_counter() - t0) * 1000.0
        print(f"findCircles: {elapsed:.3f}ms")

        return markers, error_info

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _timed_print(name: str, t0: float) -> float:
        """Print elapsed milliseconds since *t0* and return a new timestamp."""
        now = time.perf_counter()
        print(f"{name}: {(now - t0) * 1000.0:.3f}ms")
        return now

    def _findCircles(
        self,
        markers: List[Marker],
        image: np.ndarray,
        smooth: bool,
        debug: bool,
    ) -> None:
        height, width = image.shape[:2]
        min_size = width / 200.0

        t = time.perf_counter()

        # --- smooth ---------------------------------------------------
        if smooth:
            smoothed = cv2.GaussianBlur(image, (3, 3), 0.0)
        else:
            smoothed = image

        # --- threshold ------------------------------------------------
        _, mask = cv2.threshold(smoothed, 40, 255, cv2.THRESH_BINARY)
        t = self._timed_print("threshold", t)

        # --- edge + contours ------------------------------------------
        edge = cv2.Canny(smoothed, 50, 150, 3)

        # OpenCV 4.x returns (contours, hierarchy)
        contours, _ = cv2.findContours(edge, cv2.RETR_LIST, cv2.CHAIN_APPROX_NONE)
        t = self._timed_print("contour", t)

        # --- Sobel gradients ------------------------------------------
        sobel_x = cv2.Sobel(image, cv2.CV_32F, 1, 0, ksize=1)
        sobel_y = cv2.Sobel(image, cv2.CV_32F, 0, 1, ksize=1)
        t = self._timed_print("sobel", t)

        # --- subpixel refinement --------------------------------------
        circles: list = []          # cv2.fitEllipse results
        subpixel_contours: list = []  # refined contours for debug

        t = time.perf_counter()

        for contour in contours:
            if len(contour) < 5:
                continue

            x, y, w, h = cv2.boundingRect(contour)
            if w <= min_size and h <= min_size:
                continue

            area = abs(cv2.contourArea(contour))
            length = cv2.arcLength(contour, True)
            # Circularity check: area/length > length / (4*pi) * 0.7
            if area / length <= length / (4.0 * math.pi) * 0.7:
                continue

            ellipse = cv2.fitEllipse(contour)
            cx = int(ellipse[0][0] + 0.5)
            cy = int(ellipse[0][1] + 0.5)

            if not (0 <= cx < width and 0 <= cy < height):
                continue
            if mask[cy, cx] != 255:
                continue

            # --- subpixel point refinement ----------------------------
            inside = True
            subpixel_pts: List[List[float]] = []

            for pt in contour:
                px, py = int(pt[0][0]), int(pt[0][1])

                if px < 2 or py < 2 or px >= width - 2 or py >= height - 2:
                    inside = False
                    break

                delta = self._subpixelFit((px, py), image, sobel_x, sobel_y)
                if delta[0] ** 2 + delta[1] ** 2 < 1.0:
                    subpixel_pts.append([px + delta[0], py + delta[1]])

            if inside and len(subpixel_pts) >= 5:
                pts_arr = np.array(subpixel_pts, dtype=np.float32)
                subpixel_contours.append(pts_arr.reshape(-1, 1, 2).astype(np.int32))
                ellipse = cv2.fitEllipse(pts_arr)
                circles.append(ellipse)

        t = self._timed_print("subpixel", t)

        # --- deduplicate overlapping circles --------------------------
        clean_circles: list = []
        processed = [False] * len(circles)

        for i in range(len(circles)):
            if processed[i]:
                continue

            # size.area() = width * height of the rotated rect
            min_area = circles[i][1][0] * circles[i][1][1]
            min_area_circle = circles[i]
            cxi, cyi = circles[i][0]

            for j in range(len(circles)):
                if i == j or processed[j]:
                    continue

                cxj, cyj = circles[j][0]
                dist = math.hypot(cxi - cxj, cyi - cyj)
                if dist < 5.0:
                    area_j = circles[j][1][0] * circles[j][1][1]
                    if area_j < min_area:
                        min_area = area_j
                        min_area_circle = circles[j]
                    processed[j] = True

            clean_circles.append(min_area_circle)
            markers.append(((min_area_circle[0][0], min_area_circle[0][1]), min_area))
            processed[i] = True

        # --- debug output ---------------------------------------------
        if debug:
            contour_img = np.zeros((height, width, 3), dtype=np.uint8)
            circle_img = np.zeros((height, width), dtype=np.uint8)

            rng = np.random.default_rng()
            for i, sp_cnt in enumerate(subpixel_contours):
                color = (
                    int(rng.integers(0, 256)),
                    int(rng.integers(0, 256)),
                    int(rng.integers(0, 256)),
                )
                cv2.drawContours(contour_img, subpixel_contours, i, color, 1)

            for c in clean_circles:
                cv2.ellipse(circle_img, c, 255, 1)

            import os as _os
            _os.makedirs("output", exist_ok=True)
            cv2.imwrite("output/mask.png", mask)
            cv2.imwrite("output/contour.png", contour_img)
            cv2.imwrite("output/circle.png", circle_img)

    # ------------------------------------------------------------------
    # Subpixel refinement (two overloads)
    # ------------------------------------------------------------------

    def _subpixelFit(
        self,
        point: Tuple[int, int],
        image: np.ndarray,
        sobel_x: np.ndarray,
        sobel_y: np.ndarray,
    ) -> Tuple[float, float]:
        """
        1-D subpixel refinement using pre-computed Sobel gradients.

        Samples a 5-pixel line through *point* — either horizontal or
        vertical depending on the local gradient direction — and computes
        a weighted centroid of the gradient magnitudes.
        """
        px, py = point
        height, width = image.shape[:2]

        diff_x = float(sobel_x[py, px])
        diff_y = float(sobel_y[py, px])

        sum_wx = 0.0
        sum_wy = 0.0
        sum_x = 0.0
        sum_y = 0.0

        def accumulate(x: int, y: int) -> None:
            nonlocal sum_x, sum_y, sum_wx, sum_wy
            sx = float(sobel_x[y, x])
            sy = float(sobel_y[y, x])
            sum_x += sx * (x - px)
            sum_y += sy * (y - py)
            sum_wx += sx
            sum_wy += sy

        angle = math.degrees(math.atan2(diff_y, diff_x))

        if (45.0 < angle < 135.0) or (-135.0 < angle < -45.0):
            # Gradient is mostly vertical → sample along vertical line
            y0 = max(0, py - 2)
            y1 = min(height, py + 3)
            for y in range(y0, y1):
                accumulate(px, y)
        else:
            # Gradient is mostly horizontal → sample along horizontal line
            x0 = max(0, px - 2)
            x1 = min(width, px + 3)
            for x in range(x0, x1):
                accumulate(x, py)

        if sum_wx == 0.0 or sum_wy == 0.0:
            return (0.0, 0.0)

        return (sum_x / sum_wx, sum_y / sum_wy)

    # ------------------------------------------------------------------
    # Alternative 2-D polynomial surface fit (not currently called)
    # ------------------------------------------------------------------

    @staticmethod
    def _subpixelFitPoly(
        point: Tuple[int, int],
        image: np.ndarray,
        diff_x: float,
        diff_y: float,
    ) -> Tuple[float, float]:
        """
        2-D subpixel refinement using a cubic polynomial surface fit
        over a 5×5 neighbourhood and solving for the zero-crossing along
        the gradient direction.

        Note: kept for parity with the C++ API but not used in the
        default detection pipeline.
        """
        px, py = point

        # Build 25×10 design matrix and 25×1 observed grey-values
        mat_b = np.empty((25, 10), dtype=np.float64)
        mat_y = np.empty((25, 1), dtype=np.float64)

        row = 0
        for ky in range(2, -3, -1):       # k =  2, 1, 0, -1, -2
            for jx in range(-2, 3):       # j = -2, -1, 0,  1,  2
                ty = float(ky)
                tx = float(jx)

                mat_b[row, 0] = 1.0
                mat_b[row, 1] = tx
                mat_b[row, 2] = ty
                mat_b[row, 3] = tx * tx
                mat_b[row, 4] = tx * ty
                mat_b[row, 5] = ty * ty
                mat_b[row, 6] = tx * tx * tx
                mat_b[row, 7] = tx * tx * ty
                mat_b[row, 8] = ty * ty * tx
                mat_b[row, 9] = ty * ty * ty
                mat_y[row, 0] = float(image[py + ky, px + jx])
                row += 1

        # Solve least-squares:  coeffs = (BᵗB)⁻¹ Bᵗ y
        coeffs, _, _, _ = np.linalg.lstsq(mat_b, mat_y, rcond=None)
        k = coeffs.flatten()  # 10 cubic-surface coefficients

        angle = math.atan2(diff_y, diff_x)
        c = math.cos(angle)
        s = math.sin(angle)

        # p = -1/3 * (quadratic terms) / (cubic terms) evaluated along
        # the gradient direction (see C++ original for derivation).
        numerator = (
            k[3] * c * c
            + k[4] * c * s
            + k[5] * s * s
        )
        denominator = (
            k[6] * c * c * c
            + k[7] * c * c * s
            + k[8] * c * s * s
            + k[9] * s * s * s
        )
        if denominator == 0.0:
            return (0.0, 0.0)

        p = (numerator / denominator) * (-1.0 / 3.0)
        return (float(p * c), float(p * s))


# ----------------------------------------------------------------------
# (Test / demo code has been extracted to tests/test_marker_detector.py)
# ----------------------------------------------------------------------

File: markerdetector.cpp
#include "markerdetector.h"
#include "clock.h"

SLSMarkerDetector::SLSMarkerDetector()
{

}

SLSMarkerDetector::~SLSMarkerDetector()
{

}

std::string SLSMarkerDetector::detectMarkers(std::vector<std::pair<glm::vec2, float>> &markers, 
	const cv::Mat &image, bool smooth, bool debug)
{
	markers.clear();
	if (image.empty())
		return "标志点图像为空！\n";
	SLSClock clock;
	clock.begin();
	std::string errorInfo;
	try
	{
		cv::Mat grayImage;
		if (image.type() == CV_8UC3)
			cv::cvtColor(image, grayImage, cv::COLOR_BGR2GRAY);
		else
			grayImage = image;
		findCircles(markers, grayImage, smooth, debug);
	}
	catch (cv::Exception &exception)
	{
		errorInfo = exception.what();
	}
	clock.end();
	clock.displayInterval("findCircles");
	return errorInfo;
}

void SLSMarkerDetector::findCircles(std::vector<std::pair<glm::vec2, float>> &markers,
	const cv::Mat &image, bool smooth, bool debug)
{
	auto width = image.cols;
	auto height = image.rows;
	cv::Mat mask;
	SLSClock clock;
	clock.begin();
	cv::Mat smoothedImage;
	if (smooth)
		cv::GaussianBlur(image, smoothedImage, cv::Size(3, 3), 0.0);
	else
		smoothedImage = image;
	threshold(smoothedImage, mask, 40, 255, cv::THRESH_BINARY);
	clock.end();
	clock.displayInterval("threshold");
	cv::Mat edge;
	Canny(smoothedImage, edge, 50, 150, 3);
	clock.begin();
	std::vector<std::vector<cv::Point>> contours;
	cv::findContours(edge, contours, cv::RETR_LIST, cv::CHAIN_APPROX_NONE);
	clock.end();
	clock.displayInterval("contour");
	clock.begin();
	cv::Mat sobelX;
	cv::Mat sobelY;
	cv::Sobel(image, sobelX, CV_32FC1, 1, 0, 1);
	cv::Sobel(image, sobelY, CV_32FC1, 0, 1, 1);
	clock.end();
	clock.displayInterval("sobel");

	std::vector<cv::RotatedRect> circles;
	std::vector<std::vector<cv::Point>> subpixelContours;
	clock.begin();
	for (const auto &contour : contours)
	{
		if (contour.size() >= 5)
		{
			auto rect = cv::boundingRect(contour);
			if (rect.width > width / 200 || rect.height > width / 200)
			{
				auto area = fabs(cv::contourArea(contour));
				auto length = cv::arcLength(contour, true);
				//应该用拟合误差
				if (area / length > length / SLS::pi / 4.0f * 0.7f)
				{
					auto circle = cv::fitEllipse(contour);
					glm::ivec2 center(circle.center.x + 0.5f, circle.center.y + 0.5f);
					if (center.x >= 0 && center.y >= 0 && center.x < width && center.y < height)
						if (mask.at<uchar>(center.y, center.x) == 255)
						{
							auto inside = true;
							std::vector<cv::Point2f> subpixelPoints;
							for (const auto &point : contour)
							{
								if (point.x < 2 || point.y < 2 ||
									point.x >= width - 2 || point.y >= height - 2)
								{
									inside = false;
									break;
								}
								else
								{
									//auto delta = subpixelFit(point, image,
									//	sobelX.at<float>(point.y, point.x), sobelY.at<float>(point.y, point.x));
									auto delta = subpixelFit(point, image, sobelX, sobelY);
									if (delta.dot(delta) < 1.0f)
										subpixelPoints.push_back(cv::Point2f(point.x, point.y) + delta);
								}
							}
							if (inside && subpixelPoints.size() >= 5)
							{
								subpixelContours.emplace_back();
								for (const auto &subpixelPoint : subpixelPoints)
									subpixelContours.back().push_back(subpixelPoint);
								//printf("(%f %f) ", circle.center.x, circle.center.y);
								circle = cv::fitEllipse(subpixelPoints);
								//printf("(%f %f)\n", circle.center.x, circle.center.y);
								circles.push_back(circle);
							}
						}
				}
			}
		}
	}
	clock.end();
	clock.displayInterval("subpixel");

	std::vector<cv::RotatedRect> cleanCircles;
	std::vector<bool> processed(circles.size(), false);
	for (auto i = 0; i < circles.size(); ++i)
	{
		if (!processed[i])
		{
			auto minArea = circles[i].size.area();
			auto minAreaCircle = circles[i];
			for (auto j = 0; j < circles.size(); ++j)
			{
				if (i != j && !processed[j] && glm::distance(glm::vec2(circles[i].center.x,
					circles[i].center.y), glm::vec2(circles[j].center.x, circles[j].center.y)) < 5.0f)
				{
					if (circles[j].size.area() < minArea)
					{
						minArea = circles[j].size.area();
						minAreaCircle = circles[j];
					}
					processed[j] = true;
				}
			}
			cleanCircles.push_back(minAreaCircle);
			markers.emplace_back(glm::vec2(minAreaCircle.center.x, minAreaCircle.center.y), 
				minArea);
			processed[i] = true;
		}
	}

	if (debug)
	{
		cv::Mat contourImage(image.rows, image.cols, CV_8UC3, cv::Scalar(0, 0, 0));
		cv::Mat circleImage(image.rows, image.cols, CV_8UC1, cv::Scalar(0));
		for (auto i = 0; i < subpixelContours.size(); ++i)
			cv::drawContours(contourImage, subpixelContours, i, cv::Scalar(rand() % 255, 
				rand() % 255, rand() % 255), 1);
		for (const auto &circle : cleanCircles)
			cv::ellipse(circleImage, circle, cv::Scalar(255));
		imwrite("mask.png", mask);
		imwrite("contour.png", contourImage);
		imwrite("circle.png", circleImage);
	}
}

cv::Point2f SLSMarkerDetector::subpixelFit(const cv::Point &point, const cv::Mat &image,
	const cv::Mat &sobelX, const cv::Mat &sobelY)
{
	auto sumWeightsX = 0.0;
	auto sumWeightsY = 0.0;
	auto sumX = 0.0;
	auto sumY = 0.0;
	auto differenceX = sobelX.at<float>(point.y, point.x);
	auto differenceY = sobelY.at<float>(point.y, point.x);
	auto computeSum = [&sumX, &sumY, &sumWeightsX, &sumWeightsY, &sobelX, &sobelY, &point]
		(const int &x, const int &y)
	{
		sumX += sobelX.at<float>(y, x) * (x - point.x);
		sumY += sobelY.at<float>(y, x) * (y - point.y);
		sumWeightsX += sobelX.at<float>(y, x);
		sumWeightsY += sobelY.at<float>(y, x);
	};
	auto angle = glm::degrees(atan2f(differenceY, differenceX));
	if ((angle > 45.0f && angle < 135.0f) || (angle > -135.0f && angle < -45.0f))
		for (auto y = point.y - 2; y <= point.y + 2; ++y)
			computeSum(point.x, y);
	else
		for (auto x = point.x - 2; x <= point.x + 2; ++x)
			computeSum(x, point.y);
	return cv::Point2f(sumX / sumWeightsX, sumY / sumWeightsY);
}

cv::Point2f SLSMarkerDetector::subpixelFit(const cv::Point &point, const cv::Mat &image,
	float differentX, float differentY)
{
	double a[5][5];//存储周边25个点的灰度值；
	cv::Mat maty(25, 1, CV_64F);

	for (int i = 2; i >= -2; i--)
		for (int j = -2; j <= 2; j++)
			a[2 - i][j + 2] = image.at<uchar>(point.y + i, point.x + j);

	int t = 0;
	for (int i = 0; i < 5; i++)
		for (int j = 0; j < 5; j++)
		{
			maty.at<double>(t, 0) = a[i][j];
			t++;
		}
	//std::cout << maty << std::endl;
	cv::Mat matb(25, 10, CV_64F);
	double tempx;
	double tempy;
	int i = 0;
	for (int k = 2; k >= -2; k--) {
		for (int j = -2; j <= 2; j++) {
			tempx = j;
			tempy = k;

			matb.at<double>(i, 0) = 1;
			matb.at<double>(i, 1) = tempx;
			matb.at<double>(i, 2) = tempy;
			matb.at<double>(i, 3) = tempx * tempx;
			matb.at<double>(i, 4) = tempx * tempy;
			matb.at<double>(i, 5) = tempy * tempy;
			matb.at<double>(i, 6) = tempx * tempx * tempx;
			matb.at<double>(i, 7) = tempx * tempx * tempy;
			matb.at<double>(i, 8) = tempy * tempy * tempx;
			matb.at<double>(i, 9) = tempy * tempy * tempy;
			i++;

		}
	}

	cv::Mat matc;
	solve(matb, maty, matc, cv::DECOMP_SVD);
                    
	 //计算梯度
	double k[10];
	for (int i = 0; i < 10; i++)
		k[i] = matc.at<double>(i, 0);

	double gradX = differentX;// k[1];
	double gradY = differentY;// k[2];
	double angle = atanf(gradY / gradX);
	//  std::cout << cc << std::endl;
	double p = ((k[3] * cos(angle) * cos(angle) + k[4] * cos(angle) * sin(angle) + k[5] * sin(angle) * sin(angle))
		/ (k[6] * cos(angle) * cos(angle) * cos(angle) + k[7] * cos(angle) * cos(angle) * sin(angle) + k[8] * cos(angle) * sin(angle) * sin(angle) + k[9] * sin(angle) * sin(angle) * sin(angle))) * (-1.0 / 3.0);
	// std::cout << "sin" << sin(angle) << std::endl;

	return cv::Point2f(p * cos(angle), p * sin(angle));
}